src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -355,19 +668,573 @@ private CategoryDataset buildDataSet(StaplerRequest req) {
+        which it failed for each testcase.
+         */
+        Map<Integer, ArrayList<Integer>> testInfo = new HashMap<Integer, ArrayList<Integer>>();
+        /*
+        Data structure for indexing all the testcases which failed in any of the previous builds.
+         */
+        Map<String, Integer> testCaseIndex = new HashMap<>();
+        /*
+        The data series storing all the build numbers corresponding to all the previous builds.
+         */
+        XYSeries xySeries = new XYSeries(0);
+        /*
+        A map storing number of test flappers at each build.
+         */
+        flapperCountToolTip = new HashMap<>();
+        /*
+        No. of previous builds in which the testcase must have flapped to consider it as test flapper.
+         */
+        int buildHistorySize = 10;
+        /*
+        Queue of the builds and testcases which failed at that build.
+         */
+        ArrayDeque<Pair<AbstractTestResultAction<?>, HashSet<Integer>>> buildHistory = new ArrayDeque<>();
+        /*
+        Map storing the build numbers at which each testcase flapped.
+         */
+        Map<Integer, ArrayDeque<AbstractTestResultAction<?>>> testsHistory = new HashMap<>();
+        for (AbstractTestResultAction<?> testResultAction = this; testResultAction != null; testResultAction = testResultAction.getPreviousResult(AbstractTestResultAction.class)) {
+            if (++count > cap) {
+                LOGGER.log(Level.FINE, "capping test trend for {0} at {1}", new Object[]{run, cap});
+                break;
+            }
+            /*
+            Calling utility method to calculate the number of test flappers in a build.
+             */
+            if (this.run.number - testResultAction.run.number + 1 > buildHistorySize) {
+                shiftBuildHistoryUtil(buildHistory, testsHistory);
+            }
+            /*
+            Set of all the testcases which failed at the given build.
+             */
+            HashSet<Integer> buildSet = new HashSet<>();
+            hudson.tasks.junit.TestResult testResult = testResultAction.loadXml();
+            /*
+            Getting set of all the failed testcases.
+             */
+            List<CaseResult> tests = testResult.getFailedTests();
+            for (CaseResult caseResult : tests) {
+                String caseName = caseResult.getFullName();
+                if (!allPackages && !caseName.startsWith(projectLevel))
+                    continue;
+                /*
+                Initializing testInfo.
+                 */
+                if (!testCaseIndex.containsKey(caseName)) {
+                    int index = testCaseIndex.size() + 1;
+                    testCaseIndex.put(caseName, index);
+                    testInfo.put(index, new ArrayList<Integer>());
+                    List<Integer> infoList = testInfo.get(index);
+                    infoList.add(0);
+                    infoList.add(0);
+                    infoList.add(-1);
```

Comment:
leave a comments. Always try to leave a comments for a very specific operation. Also append a list instead of adding values one by one.

Reply:
Corrected.